tsdb/tsi1/report.go

```diff
@@ -0,0 +1,375 @@
+	}
+
+	return summary, nil
+}
+
+type cardinality struct {
+	name  []byte
+	short []uint32
+	set   *tsdb.SeriesIDSet
+}
+
+func (c *cardinality) add(x uint64) {
+	if c.set != nil {
+		c.set.AddNoLock(tsdb.NewSeriesID(x))
+		return
+	}
+
+	c.short = append(c.short, uint32(x)) // Series IDs never get beyond 2^32
+
+	// Cheaper to store in bitmap.
+	if len(c.short) > useBitmapN {
+		c.set = tsdb.NewSeriesIDSet()
+		for i := 0; i < len(c.short); i++ {
+			c.set.AddNoLock(tsdb.NewSeriesID(uint64(c.short[i])))
+		}
+		c.short = nil
+		return
+	}
+}
+
+func (c *cardinality) cardinality() int64 {
+	if c == nil || (c.short == nil && c.set == nil) {
+		return 0
+	}
+
+	if c.short != nil {
+		return int64(len(c.short))
+	}
+	return int64(c.set.Cardinality())
+}
+
+func (report *ReportCommand) calculateCardinalities() error {
+	itr, err := report.indexFile.MeasurementIterator()
+	if err != nil {
+		return err
+	} else if itr == nil {
+		return nil
+	}
+	defer itr.Close()
+
+	for {
+		name, err := itr.Next()
+		if err != nil {
+			return err
+		} else if name == nil {
+			return nil
+		}
+
+		if err = report.calculateMeasurementCardinalities(name); err != nil {
+			return err
+		}
+	}
+}
+
+func (report *ReportCommand) calculateMeasurementCardinalities(name []byte) error {
+	// decode org and bucket from measurement name
+	var a [16]byte
+	copy(a[:], name[:16])
+	org, bucket := tsdb.DecodeName(a)
+	if report.OrgID != nil && *report.OrgID != org ||
+		report.BucketID != nil && *report.BucketID != bucket {
+		return nil
+	}
+
+	idx := report.indexFile
+	sitr, err := idx.MeasurementSeriesIDIterator(name)
+	if err != nil {
+		return err
+	} else if sitr == nil {
+		return nil
+	}
+
+	defer sitr.Close()
+
+	var bucketCard *cardinality
+
+	// initialize map of bucket to measurements
+	if _, ok := report.byBucketMeasurement[bucket]; !ok {
+		report.byBucketMeasurement[bucket] = make(map[string]*cardinality)
+	}
+
+	if _, ok := report.byOrgBucket[org]; !ok {
+		report.byOrgBucket[org] = make(map[influxdb.ID]*cardinality)
+	}
+
+	// initialize total cardinality tracking struct for this bucket
+	if c, ok := report.byOrgBucket[org][bucket]; !ok {
+		bucketCard = &cardinality{name: []byte(bucket.String())}
+		report.byOrgBucket[org][bucket] = bucketCard
+	} else {
+		bucketCard = c
+	}
+
+	for {
+		e, err := sitr.Next()
+		if err != nil {
+			return err
+		} else if e.SeriesID.ID == 0 {
+			break
+		}
+
+		id := e.SeriesID.ID
+		if id > math.MaxUint32 {
+			return fmt.Errorf("series ID is too large: %d (max %d). Corrupted series file?", e.SeriesID, uint32(math.MaxUint32))
+		}
+
+		// add cardinality to bucket
+		bucketCard.add(id)
+
+		// retrieve tags associated with series id so we can get
+		// associated measurement
+		_, tags := report.sfile.Series(e.SeriesID)
+		if len(tags) == 0 {
+			return fmt.Errorf("series ID has empty key: %d", e.SeriesID)
+		}
+
+		// measurement name should be first tag
+		mName := string(tags[0].Value)
```

Comment:
This is right, but let's verify by checking `bytes.Equal(tags[0].Key, models.MeasurementTagKeyBytes)` and returning an error if that's not the case?